managed/services/agents/credentials_source_agent_invoker.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package agents
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/percona/pmm/api/agentpb"
+	"github.com/percona/pmm/api/inventorypb"
+	"github.com/percona/pmm/managed/models"
+	"github.com/percona/pmm/managed/utils/logger"
+)
+
+// CredentialsSourceAgentInvoker requests from agent to parse credentials-source passed in request.
+type CredentialsSourceAgentInvoker struct {
+	r *Registry
+}
+
+// NewCredentialsSourceAgentInvoker creates new CredentialsSourceAgentInvoker request.
+func NewCredentialsSourceAgentInvoker(r *Registry) *CredentialsSourceAgentInvoker {
+	return &CredentialsSourceAgentInvoker{
+		r: r,
+	}
+}
+
+// InvokeAgent sends request (with file path) to pmm-agent to parse credentials-source file.
+func (p *CredentialsSourceAgentInvoker) InvokeAgent(ctx context.Context, pmmAgentID, filePath string, serviceType models.ServiceType) (*models.CredentialsSourceParsingResult, error) {
+	l := logger.Get(ctx)
+
+	pmmAgent, err := p.r.get(pmmAgentID)
+	if err != nil {
+		return nil, err
+	}
+
+	start := time.Now()
+	defer func() {
+		if dur := time.Since(start); dur > 5*time.Second {
+			l.Warnf("Invoking agent took %s.", dur)
+		}
+	}()
+
+	request, err := createRequest(filePath, serviceType)
+	if err != nil {
+		l.Debugf("can't create ParseCredentialsSourceRequest %s", err)
+		return nil, err
+	}
+
+	resp, err := pmmAgent.channel.SendAndWaitResponse(request)
+	if err != nil {
+		return nil, err
+	}
+
+	l.Infof("ParseCredentialsSource response from agent: %+v.", resp)
+	parserResponse, ok := resp.(*agentpb.ParseCredentialsSourceResponse)
+	if !ok {
+		return nil, errors.New("wrong response from agent (not ParseCredentialsSourceResponse model)")
+	}
+	if parserResponse.Error != "" {
+		return nil, errors.New(parserResponse.Error)
+	}
+
+	return &models.CredentialsSourceParsingResult{
+		Username:      parserResponse.Username,
+		Password:      parserResponse.Password,
+		AgentPassword: parserResponse.AgentPassword,
+		Host:          parserResponse.Host,
+		Port:          parserResponse.Port,
+		Socket:        parserResponse.Socket,
+	}, nil
+}
+
+func createRequest(configPath string, serviceType models.ServiceType) (*agentpb.ParseCredentialsSourceRequest, error) {
```

Comment:
as far as I see here only service type is different here, so we could use the same mechanism like we use in https://github.com/percona/pmm/blob/4c0aaa7cea4444ae37414ce2c43b21fd1c4772b1/managed/services/management/service.go#L31 to convert from `model` to `inventorypb`

Reply:
Done.